✨ Implement single- and two-qubit gate decomposition pass

taminob opened the pull request:
## Description

This PR introduces a decomposition pass that is used as an optimization to shorten a one- or two-qubit gate series if possible.

Related to #1122 

## Checklist:

<!---
This checklist serves as a reminder of a couple of things that ensure your pull request will be merged swiftly.
-->

- [x] The pull request only contains commits that are focused and relevant to this change.
- [x] I have added appropriate tests that cover the new/changed functionality.
- [ ] I have updated the documentation to reflect these changes.
- [x] I have added entries to the changelog for any noteworthy additions, changes, fixes, or removals.
- [ ] I have added migration instructions to the upgrade guide (if needed).
- [ ] The changes follow the project's style guidelines and introduce no new warnings.
- [x] The changes are fully tested and pass the CI checks.
- [x] I have reviewed my own code changes.

taminob committed:
snapshot

taminob committed:
copied rust code

pre-commit-ci[bot] committed:
🎨 pre-commit fixes

taminob committed:
snapshot

taminob committed:
snapshot

taminob committed:
snapshot

taminob committed:
minor additions

taminob committed:
add pass to Passes.td

taminob committed:
snapshot

taminob committed:
compiling snapshot

taminob committed:
tmp

taminob committed:
switch to long double

taminob committed:
testing a.cpp

taminob committed:
add zgemm for testing, start euler decomposition implementation

taminob committed:
non-working but finished snapshot

taminob committed:
fix out-of-bounds issues

taminob committed:
test using Armadillo

taminob committed:
minor fix

taminob committed:
tmp

taminob committed:
work on gate application

taminob committed:
use laughing-umbrella implementation for decompose_two_qubit_product_gate

taminob committed:
add more eigen decomposition implementations

taminob committed:
work on applySeries()

taminob committed:
Revert "use laughing-umbrella implementation for decompose_two_qubit_product_gate"

This reverts commit 461319afefb439ace9b91ed31a88618f03165640.

taminob committed:
fix transpose_conjugate

taminob committed:
add test

taminob committed:
actually kind of working snapshot?

taminob committed:
fix test case

taminob committed:
minor fix

taminob committed:
fix test case

taminob committed:
add new test, fix tests

taminob committed:
iniital complexSeries test

taminob committed:
use Eigen for multiply() and determinant()

taminob committed:
fix gate matrix multiplication

taminob committed:
add simple (too simple) series complexity

taminob committed:
fix insertion when operations which are not part of the series are involved

taminob committed:
add gate matrices

taminob committed:
sequence complexity()

taminob committed:
print debugging :)

taminob committed:
longer complex test to pass sequence complexity check

taminob committed:
use CXGate() definition for decomposer

taminob committed:
add cnot test other direction

taminob committed:
tmp (to be deleted)

taminob committed:
re-write using Eigen

taminob committed:
delete pulse-optimizer code (KEEP for future; should be re-added at some point)

taminob committed:
use designated initializer lists

taminob committed:
fix warnings, rename all variables

taminob committed:
test normalized magic basis, allow complex eigenvector decomposition

taminob committed:
finally figured some stuff out

taminob committed:
try out lauging-umbrella weyl-chamber

taminob committed:
try to implement weylchamber.c1c2c3 algorithm

taminob committed:
on_gates two-qubit product decompose

taminob committed:
minor additions

taminob committed:
try ordering all four eigenvectors

taminob committed:
undo ordering of all columns

taminob committed:
implement algorithm based on quantumflow.canonical_decomposition

taminob committed:
sanity checks and stuff

taminob committed:
add original implementation again which seems to kind of work?

taminob committed:
generate openqasm code of generated sequence

taminob committed:
some more small fixes, maybe breakthrough?

taminob committed:
rework getTwoQubitSeries logic

taminob committed:
generate openqasm code of found series

taminob committed:
fix new getTwoQubitSeries logic

taminob committed:
working except global phase?

taminob committed:
fix global phase issues

taminob committed:
first working snapshot

taminob committed:
very minor cleanup

taminob committed:
add more sanity checks for unitary

taminob committed:
find and fix minor formula mistake

taminob committed:
test update

taminob committed:
make calculations exact without approximation

taminob committed:
respect gphase of circuit and fix series collection bug

taminob committed:
change global phase to local phase

taminob committed:
revert to global phase, fix controlled gate qubit order

taminob committed:
fix first couple of tests

taminob committed:
start working on basis set and remove dead code

taminob committed:
decompose for all bases, choose best

taminob committed:
add checks for most tests

taminob committed:
TMP, needs to be removed; allow OpenQASM export with round-trip pass

taminob committed:
add single-qubit backtracking

taminob committed:
fix atol issues

taminob committed:
add more euler bases to pattern

taminob committed:
minor cleanup (renaming, comments, disable printing)

taminob committed:
Revert "TMP, needs to be removed; allow OpenQASM export with round-trip pass"

This reverts commit 4b54f872a7b934f9d1ed5ae1fe528bcf5af243e3.

taminob committed:
remove armadillo in cmake

taminob committed:
more cleanup

taminob committed:
fix (?) tests

taminob committed:
fix final test

taminob committed:
another minor cleanup

taminob committed:
clean sanity check prints up

taminob committed:
properly add Eigen lib to cmake

taminob committed:
remove unnecessary correction

pre-commit-ci[bot] committed:
🎨 pre-commit fixes

taminob committed:
fix eigen compilation issues

taminob committed:
more cleanup

taminob committed:
clean up weyl decomposition creation

taminob committed:
remove openqasm code generation

taminob committed:
fix includes and use SmallVector instead of std::vector

taminob committed:
fix or silence linter

taminob commented:
@coderabbitai review

taminob committed:
incorporate coderabbitai feedback

flowerthrower commented on mlir/lib/Dialect/MQTOpt/Transforms/Helpers.h:
Please add a little more context to these helpers. Doesn't have to be overly verbose, but a few comments here and there (especially when it comes to recursive things) would not hurt.

taminob replied:
They are all "to be removed"

flowerthrower commented on mlir/lib/Dialect/MQTOpt/Transforms/GateDecompositionPattern.cpp:
```diff
@@ -0,0 +1,2063 @@
+        fp randA{};
+        fp randB{};
+        // For debugging the algorithm use the same RNG values from the
+        // previous Python implementation for the first random trial.
```
In our context here it should probably reference the `previous Rust implementation`; but for the same reason I like that you kept the original values here.

flowerthrower commented on mlir/lib/Dialect/MQTOpt/Transforms/GateDecompositionPattern.cpp:
```diff
@@ -0,0 +1,2063 @@
+        //
+        // This gate binds 3 parameters, we make it canonical by setting:
+        //
+        // :math:`K2_l = Id`
```
Can we make this look nicer?

taminob replied:
You both are kind of right, it would look nice in the rendered documentation, but this probably won't be included in the generated documentation, right? I'd still keep it as-is because I think it is more readable and portable than copied together UTF-8 symbols or what do you think?

The formatting got messed up by auto-wrapped comments of the auto-formatter - I'll fix it

flowerthrower reviewed (commented):
Thank you @taminob for all your work. The math and logic appear to be correctly implemented. I loosely checked the expected K1, K2, ... decomposition values with a small debug script on my Mac and was able to reproduce (roughtly) the same values as the Qiskit decomposition.  

The three remaining points are:  
1. Set up a proper test to automate exactly this verification process.  
2. Improve comments — just because the Rust implementation is poorly documented does not mean we should do the same.  
   Especially since parts of your code are already nicely commented, it makes sense to extend this good style to the copied code.
3. Revisit MLIR logic after dialect revamp

taminob commented on mlir/lib/Dialect/MQTOpt/Transforms/Helpers.h:
```diff
@@ -0,0 +1,202 @@
+  return std::nullopt;
+}
+
+inline std::optional<fp> mlirValueToFp(mlir::Value value) {
```
They are to be removed - sorry that you spent time reviewing them. The new `UnitaryInterface` will allow to directly get the unitary matrix and thus retrieving the parameter as a float becomes unnecessary.

burgholzer commented on mlir/test/Dialect/MQTOpt/Transforms/gate-decomposition.mlir:
```diff
@@ -0,0 +1,536 @@
+
+    return
+  }
+}
```
I have a couple of thoughts on this.
I agree with @flowerthrower that we can write tests for this part of the functionality that are independent from MLIR entirely. This might even justify moving the respective code outside of the MLIR subtree, but I probably wouldn't go as far.
Writing dedicated googletest-based unit tests seems to be a reasonable suggestion though that should allow us to write less brittle tests.
A great solution for equivalence checking would be our QCEC tool; but given how it builds directly on MQT Core itself, integrating it here (especially in C++) would be a mess.
Given how we are talking about 1- and 2- qubit gates only, it should be entirely feasible to write a very small script that multiplied the matrices and either does the equivalent of `np.all_close()` or computes `tr(U^dagger V)/2^n` if `U` and `V` are the two matrices to be compared and `n` is the number of qubits. Then, one would assert that the result is close enough to 1.

It would be good to come up with a couple of unit tests that assert the implementation produces sequences of gates we would expect and that this behaviour is maintained across updates.
I'll leave it at that for now. I'll probably have further comments when I actually look through the PR itself.

taminob committed:
remove global phase accumulation

taminob committed:
fix :math: latex formatting

taminob committed:
minor comment fixes

taminob committed:
replace CHECK-DAG by CHECK again

taminob committed:
add comments to weyl decomposition

taminob committed:
extract TwoQubitWeylDecomposition::getTrace()

taminob committed:
transform basis decomposer into actual class with private members

taminob committed:
attempt to fix CI test

taminob committed:
add supercontrolled check

taminob committed:
extract decomposition into headers

taminob committed:
slight improvement to weyl decomposition for testing purposes

taminob committed:
add unittest for weyl decomposition

taminob committed:
fix linter issues

taminob committed:
add tests for basis decomposer

taminob committed:
fix linter issues

taminob committed:
clang-tidy: Add IgnoreHeader for include cleaner for Eigen

taminob committed:
remove exact values from MLIR test

taminob committed:
fix mlir test/linter issues

taminob committed:
remove std::numbers because not supported by all compilers in CI

llvm::numbers could be used instead

taminob committed:
remove unused Helpers.h from src dir

taminob committed:
remove auto* because of windows build

burgholzer commented on mlir/lib/Dialect/MQTOpt/Transforms/GateDecompositionPattern.cpp:
```diff
@@ -0,0 +1,2061 @@
+          {.op = nextGate, .qubitIds = {firstQubitId, secondQubitId}});
+      complexity += getComplexity(helpers::getQcType(nextGate), 2);
+      return true;
+    }
```
Hm. I remember stumbling over this as well. I can't quite remover where, but it could have been as part of the dialect rewrite. I think that I added a NOLINT for that single occurrence.
Stupid thought, but have you tried `auto&`? do any of the systems complain about that?
if there are multiple instances of the rule violation in one place, you can always sandwich them between `// NOLINTBEGIN(...)` and `// NOLINTEND(...)` to hide the warning in a moderately clean way.

burgholzer commented:
> I reworked the tests now (added unittests and made the MLIR tests more robust/less precise).
> 
> A couple of questions:
> 
> * Is this approach to testing good now or do you think the MLIR tests are too imprecise?

I just skimmed through both the lit tests and the unit tests and I think the overall direction is fine. I am pretty sure that I will have some more detailed comments when I check the PR in more detail.

> * Should I adjust this PR to also include the single-qubit decomposition in the pattern or should we do this in a follow-up PR? The full code already exists and it would be the most performant way to add the single-qubit decomposition in the same pass because the qubit series is collected anyway (although the pattern is quite big aleady).

Hm. multiple thoughts on this.
I foresee two different use cases for the synthesis routines here:
- collection and resynthesis of arbitrary runs of two-qubit blocks with no real constraints on the gate set (maybe with a parameter that let's one choose the entangling gate; like CX or CZ or RZZ or ...)
- synthesis of two-qubit operations (or blocks thereof) to a dedicated/native gateset of a divide (where only those operations are permitted in the decomposition).

These two might actually be more or less the same, the first might just be the second with some reasonable and relaxed defaults on the native gate-set.
I think it could make sense to include the single-qubit synthesis here as well because, as you said, the series are collected anyway, so one might as well apply this to those sequences in the same go.
The only concern that I would probably have is performance. I would assume that single-qubit synthesis is much more efficient than the two-qubit one. If all one does is two-qubit synthesis, then this might become costly at some point.

> * The Windows ARM build has an internal alignment issue in Eigen (https://libeigen.gitlab.io/eigen/docs-nightly/group__TopicUnalignedArrayAssert.html; the link in the build log is broken). Will you resolve this issue already in the dialect rewrite that'll introduce Eigen?

I am not quite sure that I follow this. That link says:
> There are 4 known causes for this issue. If you can target [c++17] only with a recent compiler (e.g., GCC>=7, clang>=5, MSVC>=19.12), then you're lucky: enabling c++17 should be enough (if not, please [report](http://eigen.tuxfamily.org/bz/) to us). Otherwise, please read on to understand those issues and learn how to fix them.

We are using more modern compilers than this (by far). The current CI runs report
> -- The C compiler identification is MSVC 19.44.35217.0
> -- The CXX compiler identification is MSVC 19.44.35217.0

This seems to indicate to me that you might need to dig a little deeper there.
I am not yet 100% sure if the first version of the dialect rewrite will immediately include the Eigen dependency. At the moment, the following signature is used in the unitary interface
```
InterfaceMethod<
            "Attempts to extract the static unitary matrix. "
            "Returns std::nullopt if the operation is symbolic or dynamic.",
            "DenseElementsAttr", "tryGetStaticMatrix", (ins)
        >,
```
An example of how that Attr is constructed would be
```
inline DenseElementsAttr getMatrixX(MLIRContext* ctx) {
  const auto& complexType = ComplexType::get(Float64Type::get(ctx));
  const auto& type = RankedTensorType::get({2, 2}, complexType);
  const auto& matrix = {0.0 + 0i, 1.0 + 0i,  // row 0
                        1.0 + 0i, 0.0 + 0i}; // row 1
  return DenseElementsAttr::get(type, matrix);
}
```
This might not be final yet; mostly because I still want to explore some options that might allow us to make the gates/gate matrices singletons and avoid repeated constructions of matrices for gates where this is definitely not necessary.

> * I currently use two `unsupported` (so experimental) headers of Eigen. Is this fine or do we want to implement that ourselves? It's for `kroneckerProduct()` and `Eigen::Matrix::exp()`

I am not too familiar with the details of `Eigen`, but just from the name alone, I have an aversion to using such headers. Is there any kind of intuition for how well these are tested/work/can be used?
Kronecker Product shouldn't be particularly hard to implement, I suppose.
The Matrix exponential certainly is a little more complicated.
This is definitely not a decision, but just a little bit of caution.

taminob committed:
try to fix windows ARM ci

taminob committed:
pass matrices by reference

taminob committed:
allow non-optimal vectorization

pre-commit-ci[bot] committed:
🎨 pre-commit fixes

taminob committed:
try to fix

taminob commented:
> I'll look into it, although it's a bit hard on a different architecture. Could well be that it is caused by differences in alignment on ARM.

Unfortunately, I couldn't really find the source of the error - also because I didn't manage to get a stacktrace in the CI and so can't really pinpoint the issue. I tried using C++23 `stacktrace` which did compile, but I didn't manage to override Eigen's `eigen_assert` at the place this is triggered. I didn't manage to reproduce this locally.

I think the cause might be one of https://libeigen.gitlab.io/eigen/docs-nightly/group__TopicPassingByValue.html or https://libeigen.gitlab.io/eigen/docs-nightly/group__TopicStlContainers.html (which should be fixed since C++11, but maybe LLVM+MSVC behaves buggy here - who knows) which are quite serious and need to be considered when using Eigen (at least I wasn't aware of it).

I'd suggest to simply set `EIGEN_MAX_STATIC_ALIGN_BYTES` to zero for ARM+Windows as advised in the documentation. The drawback is potentially non-optimal vectorization. I did this by defining `EIGEN_DONT_ALIGN_STATICALLY`.

taminob committed:
integrate single-qubit decomposition

taminob committed:
add tests for single-qubit decomposition

taminob committed:
add unittests for euler decomposition

taminob committed:
fix linter issues

taminob committed:
fix include cleaner issues

taminob committed:
add test parametrizations for other specializations

taminob committed:
add more matrix definitions

taminob committed:
handle barrier gates in series collection

pre-commit-ci[bot] committed:
🎨 pre-commit fixes

taminob committed:
fix include cleaner issues

taminob committed:
remove unused globalPhase from TwoQubitSeries

taminob committed:
fix barrier handling at start of series

taminob committed:
simplify weyl specialization test parameters

taminob committed:
remove dynamic_casts due to RTTI issue

taminob committed:
Revert "remove dynamic_casts due to RTTI issue"

This reverts commit dc13927d78c687e4ab0d22adcef0724ada281a52.

taminob committed:
move GateDecompositionPass to new QCO dialect

taminob committed:
add GateDecompositionPass to compiler pipeline

pre-commit-ci[bot] committed:
🎨 pre-commit fixes

taminob committed:
add UnitaryMatrixType to operation and add Eigen dependency to dialect

taminob committed:
first non-working attempt

taminob committed:
second non-working attempt

taminob committed:
compiling approach

taminob committed:
formatting

taminob committed:
avoid use of implementation-dependent argument names

taminob committed:
finally got template function working

taminob committed:
add to compiler pipeline and add (failing) test case

taminob committed:
make unitary function const, remove HasMatrixDefinition bool template parameter

taminob committed:
start working on ctrl matrix

taminob committed:
use new unitary matrix functions in GateDecompositionPattern

taminob committed:
add debug prints

pre-commit-ci[bot] committed:
🎨 pre-commit fixes

taminob committed:
add helpers for CtrlOp matrix

taminob committed:
add matrix definition to all operations

taminob committed:
fixes for unitary matrix definition

taminob committed:
fixes for gate decomposition pattern

taminob committed:
fix formatting in MatrixUtils.h

taminob committed:
(dirty) fix for CtrlOp in gate decomposition pattern; runs, but SSA error at end and wrong result

taminob committed:
Merge remote-tracking branch 'origin/main' into taminob/mlir-two-qubit-gate-decomposition-pass

taminob committed:
:memo: Add PR of unitary matrices to CHANGELOG

taminob committed:
remove matrix draft remains

taminob committed:
fix compilation issues with new matrix implementation

pre-commit-ci[bot] committed:
🎨 pre-commit fixes

taminob committed:
fix issue with CtrlOp

taminob committed:
fixes and improvements (forceApplication+singleQubitOnly) for QCO dialect

taminob committed:
Merge remote-tracking branch 'origin/main' into taminob/mlir-two-qubit-gate-decomposition-pass

pre-commit-ci[bot] committed:
🎨 pre-commit fixes

taminob committed:
fix nested ops issue and others in pass/pattern

taminob committed:
set GPhase complexity to 0 and fix complexity calculation for backtracking

taminob committed:
thesis evaluation: add statistics and disable canonicalizations

taminob committed:
evaluation scripts

taminob committed:
evaluation results

taminob committed:
evaluation script fixes and improvements

taminob committed:
update evaluation results

taminob committed:
evaluate circuit collection for qiskit, total_evaluate fixes

taminob committed:
update evaluation results

taminob committed:
evaluation script improvements

taminob committed:
update evaluation figures

taminob committed:
Revert thesis evaluation changes

This reverts the following commits:

* c66c4ec4dc7a8d16b981d0f80acc592355745109
* 007b575f7d2d3d7a946ed207d794620a3268585c
* 2ddfc6cad2bef4ac19d614fa36bde3efde8d9e66
* 5e676bf30a84e872e0193066e551e67687c9d450
* b4af378f093b69e4091dff808c6eaaccee6a90e4
* 3a8b02a3eca3dc07a00d0dd82f3aaf19ecf7bd1b
* 90c46f47be72c56eb26afb22b44d96fa60e0d4d4
* a79324cc049d189cf817a5ccadf723e2c5e72c15
* c46f15970399ac5932089cebe4a1a77d2476ba57

taminob committed:
Merge remote-tracking branch 'origin/main' into taminob/mlir-two-qubit-gate-decomposition-pass

taminob committed:
:memo: Add PR to CHANGELOG

taminob committed:
fix typo in comment

taminob committed:
improve naming in getDefaultNbasis

taminob committed:
coderabbit feedback and cleanup

taminob committed:
remove broken unittest; need to think of ways to verify optimizations

taminob commented on mlir/include/mlir/Passes/Decomposition/EulerBasis.h:
```diff
@@ -0,0 +1,76 @@
+  }
+  throw std::invalid_argument{
+      "Unsupported euler basis for translation to gate types"};
+}
```
Should be fixed now, right?

taminob committed:
remove helpers::kroneckerProduct, IDENTITY_GATE and fix unitary matrices

taminob committed:
more fixes

taminob committed:
relax isUnitaryMatrix() default tolerance to Eigen's default value

taminob committed:
fix mysterious segfault in Eigen::kroneckerProduct (still no idea why...)

taminob committed:
reduce random test time limit to 3s

This should still be sufficient for proper testing but will reduce the
amount of wasted CPU and development time when not necessary. When doing
active development on the decompositions, this can still be manually
increased again.

taminob committed:
add C_ prefix to complex IM, M_IM

taminob committed:
remove custom types

taminob committed:
remove constants and other stuff

taminob committed:
fix typo in comment

taminob commented on mlir/include/mlir/Passes/Decomposition/BasisDecomposer.h:
```diff
@@ -0,0 +1,551 @@
+      return bestIndex;
+    };
+    // number of basis gates that need to be used in the decomposition
+    auto bestNbasis = numBasisGateUses.value_or(getDefaultNbasis());
```
I will replace min() by lowest(), but will keep the -1 for bestIndex since that will fail pretty soon in `chooseDecomposition()`. Using `std::optional` would make the code more complicated for little benefit

taminob committed:
replace exceptions by llvm error reporting

taminob committed:
some coderabbitai comments and complexity caching

taminob committed:
remove crash test

taminob committed:
make Weyl decomposition non-trivial class; some other fixes

taminob committed:
better random basis tests and other fixes

taminob committed:
more coderabbitai fixes

taminob committed:
add utils for tests

taminob commented on mlir/include/mlir/Passes/Decomposition/GateSequence.h:
```diff
@@ -0,0 +1,94 @@
+      c += helpers::getComplexity(qc::GPhase, 0);
+    }
+    return c;
+  }
```
This is intentional because it should be possible to modify the complexity without an update in the rest of the code

taminob committed:
coderabbit feedback

taminob committed:
fix includes

taminob committed:
separate definitions into source files

pre-commit-ci[bot] committed:
🎨 pre-commit fixes

taminob committed:
incorporate coderabbitai feedback and fix linter issues

taminob committed:
remove constexpr from gate constants

pre-commit-ci[bot] committed:
🎨 pre-commit fixes

taminob commented on mlir/lib/Passes/Patterns/Decomposition/BasisDecomposer.cpp:
```diff
@@ -0,0 +1,401 @@
+    return bestIndex;
+  };
+  // number of basis gates that need to be used in the decomposition
+  auto bestNbasis = numBasisGateUses.value_or(getDefaultNbasis());
```
added comment and explicit casting

taminob committed:
improve comments, other minor fixes

taminob committed:
add some caching for available gates and improve comments

taminob committed:
Merge remote-tracking branch 'origin/main' into taminob/mlir-two-qubit-gate-decomposition-pass

taminob committed:
minor comment and one const fix

taminob committed:
fix annoying static initialization issue in gtest with Eigen

taminob committed:
switch to fully deterministic random matrix generation

taminob committed:
fix potential alignment issues in SmallVector inline storage

taminob committed:
fix alignment issue for static-sized matrices on stack

Stack alignment is only 16 bytes, but AVX instructions required 32 byte
alignment. Can be fixed by making all stack-variables aligned, but
lowering the alignment (and thus limiting the optimizations applied)
seems like the more long-term solution - especially considering that
this issue is not detected in the CI.

pre-commit-ci[bot] committed:
🎨 pre-commit fixes

taminob committed:
Merge remote-tracking branch 'origin/main' into taminob/mlir-two-qubit-gate-decomposition-pass

taminob committed:
Merge remote-tracking branch 'origin/main' into taminob/mlir-two-qubit-gate-decomposition-pass

taminob committed:
linter/coderabbit fixes

taminob committed:
remove static random generator for better reproducibility

taminob committed:
minor fixes

pre-commit-ci[bot] committed:
🎨 pre-commit fixes

taminob committed:
minor fixes

taminob commented on mlir/lib/Dialect/QCO/IR/CMakeLists.txt:
```diff
@@ -49,3 +49,6 @@ target_sources(
          ${IR_HEADERS_BUILD})
+
+# fix alignment issues for stack-allocated fixed-size matrices
+target_compile_definitions(MLIRQCODialect PUBLIC EIGEN_MAX_STATIC_ALIGN_BYTES=16)
```
I experienced the segfaults with Linux on x86_64 architecture. The CI probably only succeeds because the machines don't support AVX instructions

taminob commented:
@burgholzer this PR should be ready for your review. I'm sorry this PR is so huge, but it contains now (as discussed before) both the logic for single- and two-qubit decomposition. To avoid bloating this PR even more (I think this also makes the coderabbit reviews worse), I did not add more extensive tests - I'd maybe suggest doing this in a follow-up in #1499 where coderabbit generated some more edge case test cases. I'm not sure about how to test it in the context of the pipeline since it will be hard to test passes in isolation once we start adding lots of optimizations to the pipeline.

I'll also mention you in all of the remaining open coderabbit comments since the history of this PR is quite long by now.

Once this PR is merged and we agree on the project structure for optimizations, I'd start working on the swap-reconstruction-and-elision pass.

taminob committed:
reapply thesis-evaluation (revert 8b7f0bc5640cd93093db534bf608652c02a694c0)

taminob committed:
evaluation script update: automated Qiskit Rust eval

taminob committed:
evaluation results

taminob committed:
update qiskit python results without Rust timing code

taminob committed:
add Qiskit copyright notice to decomposition classes

taminob committed:
add XZX euler basis to mqt decomposition

taminob committed:
update evaluation results for mqt and figures

taminob committed:
fix evaluation graph y axis + add improved circuit calculation

taminob committed:
update figures for corrected y axis

taminob committed:
Revert thesis-evaluation changes

This reverts the following commits:

* b21868f337a1dc675959939cab7fb9ee0880c6c4
* f4b6d7fc865a9fab6e7f67733d82bd64fa77c25f
* 0b9a10d6c3d99c7138878e549d3dda205acc691d
* e61854f57fd096e9bd7f9629086fcec55c327dac
* 97755cf753f9ecea468ccbb54923f982346b1e5a
* 8f58cb401706e3f457de7e3f148c8c33b7bf3063
* 310d35ab13ca7851d6d3b119a24626b547b0f93f

MatthiasReumann commented on mlir/include/mlir/Passes/Passes.td:
```diff
@@ -0,0 +1,27 @@
+// Optimization Passes
+//===----------------------------------------------------------------------===//
+
+def GateDecompositionPass : Pass<"gate-decomposition", "mlir::ModuleOp"> {
```
I have a very stupid question: Why is gate decomposition considered an optimization pass?

taminob replied:
No, that is a completely fair question. The reason is mainly that my master's thesis is about the implementation of optimization passes - in the beginning, I didn't even know that the decompositions are usually not used for optimizing circuits. So the current pass is structured as an optimization which only decpmposes if the decomposed circuit is less complex than before. I already added some configuration options to the pattern with the native gate synthesis in mind, but that is for future expansion.

taminob committed:
Merge remote-tracking branch 'origin/main' into taminob/mlir-two-qubit-gate-decomposition-pass

taminob committed:
fix compilation issues after merge

taminob committed:
Merge remote-tracking branch 'origin/main' into taminob/mlir-two-qubit-gate-decomposition-pass

taminob committed:
fix linter issue

taminob committed:
remove eigen alignment limit from cmake

MatthiasReumann committed:
Merge branch 'main' into HEAD

taminob committed:
cleanup [skip ci]

burgholzer committed:
Merge branch 'main' into taminob/mlir-two-qubit-gate-decomposition-pass

Signed-off-by: burgholzer <burgholzer@me.com>

# Conflicts:
#	CHANGELOG.md
#	mlir/lib/Compiler/CompilerPipeline.cpp
#	mlir/lib/Dialect/QCO/Transforms/CMakeLists.txt

burgholzer commented:
I am closing this PR for now without deleting the underlying branch.
This has been a very valuable first step towards implementing the single- and two-qubit decomposition logic.
However, it has just grown too large to bring to a mergeable state here.
The implementation will continue in follow-up PRs (such as #1672 or #1665).
